### PR TITLE
[FEATURE] Add Rector

### LINKED ISSUE #125: Evaluate typo3-rector

### PR DESCRIPTION
**Pull request by Starkmann**
This part of #851.

In later parts, we'll add a Rector wrapper to `runTests.sh` and a CI job, and apply possible changes suggested by Rector.

Part of #851
Closes #125

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -60,6 +60,7 @@
 		"saschaegerer/phpstan-typo3": "^1.9.1",
 		"seld/jsonlint": "^1.10.0",
 		"squizlabs/php_codesniffer": "^3.7.2",
+		"ssch/typo3-rector": "^1.3",
```
Comment (oliverklee) [inline composer.json:63]: For dev dependencies, please always let's use the full version here (and also the latest as far as possible), i.e. `^1.5.2`.

File: rector.php [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+use Ssch\TYPO3Rector\Set\Typo3SetList;
+
+return static function (RectorConfig $rectorConfig): void {
+
+    // If you want to override the number of spaces for your typoscript files you can define it here, the default value is 4
+    // $parameters = $rectorConfig->parameters();
+    // $parameters->set(Typo3Option::TYPOSCRIPT_INDENT_SIZE, 2);
+
+    $rectorConfig->sets([
+        Typo3LevelSetList::UP_TO_TYPO3_11,
```
Comment (oliverklee) [inline rector.php:22]: I suggest also adding `LevelSetList::UP_TO_PHP_74`.

**Review by oliverklee — CHANGES_REQUESTED** [review]
We also should document somewhere (e.g., in the manual) how to run Rector via DDEV and `runTests.sh`.

Also, we should add the `rector.php` to the list of not-packaged files in `.gitattributes` and the `prepare-release`  Composer script.

**Comment by lukaszuznanski** [comment]
@Starkmann are you willing to work on this? If so, I will close my #992

File: rector.php (hunk 2 above)
Comment (simonschaufi) [reply rector.php:22]: I have also added this line to the rector template

**Comment by lukaszuznanski** [comment]
Hi,

We've implemented a policy to close pull requests that have been inactive for over six months. If you're still interested in contributing and would like to continue working on this pull request, we're more than happy to keep it open.

Please take a moment to review your pull request and let us know if you'd like to proceed with it. If we don't receive a response within the next 30 days, we'll assume that you no longer wish to pursue it, and we'll proceed with closing the pull request.

Feel free to reach out if you have any questions or need further assistance.

File: rector.php (hunk 2 above)
Comment (oliverklee) [reply rector.php]: Done.

File: composer.json (hunk 1 above)
Comment (oliverklee) [reply composer.json]: Done.

File: rector.php [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+    ->withPhpVersion(PhpVersion::PHP_74)
+    ->withRules([
+        ConvertImplicitVariablesToExplicitGlobalsRector::class,
+    ])
+    // If you use importNames(), you should consider excluding some TYPO3 files.
+    ->withSkip([
+        // @see https://github.com/sabbelasichon/typo3-rector/issues/2536
+        __DIR__ . '/.github/*',
+        __DIR__ . '/.Build/*',
+        NameImportingPostRector::class => [
```
Comment (simonschaufi) [inline rector.php:88]: this one can be removed as it is obsolete with typo3 11

File: rector.php [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    ])
+    // To have a better analysis from PHPStan, we teach it here some more things
+    ->withPHPStanConfigs([
+        Typo3Option::PHPSTAN_FOR_RECTOR_PATH,
+    ])
+    ->withPhpVersion(PhpVersion::PHP_74)
+    ->withRules([
+        ConvertImplicitVariablesToExplicitGlobalsRector::class,
+    ])
+    // If you use importNames(), you should consider excluding some TYPO3 files.
```
[the remark this answers is not in the document]
Comment (oliverklee) [reply rector.php:82]: I still don't quite know what you mean. Would you be willing to suggest the wording in a review comment?

[the remark this answers is not in the document]
Comment (simonschaufi) [reply rector.php:82]: in the old config it looks like this https://github.com/simonschaufi/gkh_rss_import/blob/main/Build/rector/rector.php#L55-L56 in the new it is with "with..."

[the remark this answers is not in the document]
Comment (oliverklee) [reply rector.php:102]: Done.

File: rector.php [hunk 5]
```diff
@@ -0,0 +1,108 @@
+<?php
```
Comment (DanielSiepmann) [inline rector.php:1]: We need to add this file to `.gitattributes` as `/rector.php export-ignore`

**Review by DanielSiepmann — CHANGES_REQUESTED** [review]

**Review by lukaszuznanski — APPROVED** [review]

**Comment by sbuerk** [comment]
Not sure if moving rector installation to a subfolder composer.json would be feasonable directly, as for TYPO3 v13 and rector there is a conflict regard the nikic php parser librady version (for the overall working forward to TYPO3 v13).

**Review by oliverklee — APPROVED** [review]

File: rector.php [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+use Ssch\TYPO3Rector\Configuration\Typo3Option;
+use Ssch\TYPO3Rector\Set\Typo3LevelSetList;
+use Ssch\TYPO3Rector\Set\Typo3SetList;
+use Ssch\Typo3RectorTestingFramework\Set\TYPO3TestingFrameworkSetList;
+
+return RectorConfig::configure()
+    ->withPaths([
+        __DIR__ . '/Classes/',
+        __DIR__ . '/Configuration/',
+        __DIR__ . '/Tests/',
```
Comment (simonschaufi) [inline rector.php:25]: __DIR__ . './ext_emconf.php',
        __DIR__ . './ext_localconf.php',

File: rector.php [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        ExtEmConfRector::PHP_VERSION_CONSTRAINT => '7.4.0-8.3.99',
+        ExtEmConfRector::TYPO3_VERSION_CONSTRAINT => '11.5.4-12.4.99',
+        ExtEmConfRector::ADDITIONAL_VALUES_TO_BE_REMOVED => [],
+    ])
+    ->withSkip([
+        // CodeQuality
+        ExplicitBoolCompareRector::class,
+        SwitchNegatedTernaryRector::class,
+        // Strict
+        DisallowedEmptyRuleFixerRector::class,
```
Comment (simonschaufi) [inline rector.php:108]: should be checked individually per project

**Comment by oliverklee** [comment]
> Not sure if moving rector installation to a subfolder composer.json would be feasonable directly,

I'd prefer to do that in a separate step.

File: rector.php [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        // PHPUnitSetList::PHPUNIT_100,
+        // PHPUnitSetList::PHPUNIT_CODE_QUALITY,
+
+        // TYPO3 Sets
+        // https://github.com/sabbelasichon/typo3-rector/blob/main/src/Set/Typo3LevelSetList.php
+        // https://github.com/sabbelasichon/typo3-rector/blob/main/src/Set/Typo3SetList.php
+
+        Typo3SetList::TYPO3_11,
+
+        // Typo3SetList::CODE_QUALITY,
```
[the remark this answers is not in the document]
Comment (oliverklee) [reply rector.php:80]: Done.

File: rector.php (hunk 7 above)
Comment (oliverklee) [reply rector.php:108]: Done.

File: rector.php (hunk 6 above)
Comment (oliverklee) [reply rector.php:25]: Done.

**Review by simonschaufi — APPROVED** [review]

**Review by oliverklee — APPROVED** [review]

**Review by lukaszuznanski — APPROVED** [review]

**Review by DanielSiepmann — APPROVED** [review]

**Pull request merged**